### ISSUE TITLE
pathlib .suffix, .suffixes, .stem unexpected behavior for pathname with trailing dot

### ISSUE THREAD

**Issue by dd5ce2ff-95a1-4e28-ace2-6d841dd59913** [python/cpython#82805]
BPO | [38624](https://bugs.python.org/issue38624)
--- | :---
Nosy | @pitrou

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2019-10-28.22:54:26.232>
labels = ['3.8', 'type-bug', 'library']
title = 'pathlib .suffix, .suffixes, .stem unexpected behavior for pathname with trailing dot'
updated_at = <Date 2019-10-29.01:59:29.319>
user = 'https://bugs.python.org/inyeollee'
```

bugs.python.org fields:
```python
activity = <Date 2019-10-29.01:59:29.319>
actor = 'xtreak'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2019-10-28.22:54:26.232>
creator = 'inyeollee'
dependencies = []
files = []
hgrepos = []
issue_num = 38624
keywords = []
message_count = 1.0
messages = ['355600']
nosy_count = 2.0
nosy_names = ['pitrou', 'inyeollee']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue38624'
versions = ['Python 3.8']
```

</p></details>

### Linked PRs
* gh-118952

**Comment by ketozhang** [comment]
In my attempt to work on this bug, I found it is not possible to make a fix without deciding whether files with trailing dots meaningfully contain a suffix. Currently, Python implicitly defines strings with trailing dots to be an invalid suffix, but does not error:

```py
# Yes, obvious case
>>> Path("foo.tar").suffix
'.tar'

# Trailing dot is not a suffix
>>> Path("foo.tar.").suffix
''

# Consistent with before, but confusingly it's not a suffix
>>> Path("foo").with_suffix(".tar.")
PosixPath('foo.tar.')
>>> PosixPath('foo.tar.').with_suffix(".tar.").suffix
''
```

So, this isn't a bug if "trailing dots is not a valid suffix". However, this rule is seemingly ignored in `with_suffix()` since we are able to pass in a string with a trailing dot.  At the same time it is not true that `p.with_suffix(s).suffix == s` since you can pass in compound suffixes.

My proposed solution (outside the scope of this issue) is to change `with_suffix` such that it raises an exception when passed a string with trailing dots.

**Comment by barneygale** [comment]
I agree with the original bug report - `Path("foo.tar.").suffix` should give you `'.'`. That would match `os.path.splitext()` behaviour.

**Comment by ketozhang** [comment]
@barneygale Good point, then it must be that

```py
>>> Path("foo.tar.").suffixes
[".tar", "."]
```

As well as we needing to remove the restriction with `with_suffix(".")`:

```
>>> Path("foo.tar").with_suffix(".")
Path("foo.tar.")
```

**Comment by barneygale** [comment]
More discussion in #100157, which I've closed as a duplicate.

**Comment by barneygale** [comment]
PR available: https://github.com/python/cpython/pull/118952

**Comment by barneygale** [comment]
Fixed in 3.14 / [e418fc3](https://github.com/python/cpython/commit/e418fc3a6e7bade68ab5dfe72f14ddba28e6acb5) / #118952
